cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java

```diff
@@ -886,33 +886,83 @@ public synchronized void userTriggeredStopExecution() {
+   * @param concurrentLeaderMovements The maximum number of concurrent leader movements
+   *                                  (if null, use num.concurrent.leader.movements).
+   * @param skipHardGoalCheck True if the provided {@code goals} do not have to contain all hard goals, false otherwise.
+   * @param replicaMovementStrategy The strategy used to determine the execution order of generated replica movement tasks
+   *                                (if null, use default.replica.movement.strategies).
+   * @param excludeRecentlyDemotedBrokers Exclude recently demoted brokers from proposal generation for leadership transfer.
+   * @param excludeRecentlyRemovedBrokers Exclude recently removed brokers from proposal generation for replica transfer.
+   * @param dryRun Whether it is a dry run or not.
+   * @param uuid UUID of the execution.
+   *
+   * @return The optimization result.
+   * @throws KafkaCruiseControlException When any exception occurred during the topic configuration updating.
    */
-  private Set<ExecutionProposal> maybeUpdateTopicReplicationFactor(Cluster cluster,
-                                                                   short replicationFactor,
-                                                                   Set<String> topics,
-                                                                   boolean skipTopicRackAwarenessCheck) {
+  public GoalOptimizer.OptimizerResult updateTopicConfiguration(Pattern topicPattern,
+                                                                List<String> goals,
+                                                                short replicationFactor,
+                                                                boolean skipTopicRackAwarenessCheck,
+                                                                ModelCompletenessRequirements requirements,
+                                                                OperationProgress operationProgress,
+                                                                boolean allowCapacityEstimation,
+                                                                Integer concurrentInterBrokerPartitionMovements,
+                                                                Integer concurrentLeaderMovements,
+                                                                boolean skipHardGoalCheck,
+                                                                ReplicaMovementStrategy replicaMovementStrategy,
+                                                                boolean excludeRecentlyDemotedBrokers,
+                                                                boolean excludeRecentlyRemovedBrokers,
+                                                                boolean dryRun,
+                                                                String uuid)
+      throws KafkaCruiseControlException {
```

Comment:
Nit: Is it possible to move some logic from within this function to smaller function(s)? This function seems quite monolithic.